### ISSUE TITLE
17/18 ST Q1c

### ISSUE THREAD

**Issue by dumbasssoul** [nus-cs2030/2021-s2#381]
how do i go about doing q1c? im not sure how to store the values of each (Xk - μ)^2 as wont the value require μ to be generated first?

**Comment by Buwoo** [comment]
The question didn't state that we are required to complete everything within a single stream pipeline, only criteria is cannot use explicit looping/recursive implementations and data structures. 
So I think its easiest (though may not be the most elegant solution) to just use the IntStream average method to get the mean, then if the mean is not empty, you can perform the necessary calculations using the mean value as generated earlier. 
Hope this helps!

**Comment by jesslynlee** [comment]
hello, this is my attempt.

`public static OptionalDouble variance(int[] data) {
	if (data.length < 2) { 
		return OptionalDouble.empty();
	}
	int n = data.length;
	double mean = Arrays.stream(data).asDoubleStream().sum() / n;
	double num = Arrays.stream(data).mapToDouble(x -> (x-mean)*(x-mean)).sum();
	return OptionalDouble.of(num/(n-1));
}`

**Comment by davidwang98** [comment]
My friends and I found the solution. Here is our working.

```
public static OptionalDouble variance(int[] data) {
    return OptionalDouble.of(IntStream.of(data)
		.mapToDouble(x -> (double) x) // convert everything to double
		.map(x -> x - IntStream.of(data).average().getAsDouble()) // x - minus average
		.map(x -> x * x) // this is to square the numerator
		.reduce(0, (x,y) -> (x + (y/(data.length - 1))))); // sum everything together
}
```

**Comment by Buwoo** [comment]
> My friends and I found the solution. Here is our working.
> 
> ```
> public static OptionalDouble variance(int[] data) {
>     return OptionalDouble.of(IntStream.of(data)
> 		.mapToDouble(x -> (double) x) // convert everything to double
> 		.map(x -> x - IntStream.of(data).average().getAsDouble()) // x - minus average
> 		.map(x -> x * x) // this is to square the numerator
> 		.reduce(0, (x,y) -> (x + (y/(data.length - 1))))); // sum everything together
> }
> ```

I don't think this method works when there are 0 elements in the data array. It will return OptionalDouble[0.0] instead of OptionalDouble.empty. I think can use an if statement to catch this corner case

**Comment by dumbasssoul** [comment]
thanks all for the help!

**Comment by danieltwh** [comment]
> > My friends and I found the solution. Here is our working.
> > ```
> > public static OptionalDouble variance(int[] data) {
> >     return OptionalDouble.of(IntStream.of(data)
> > 		.mapToDouble(x -> (double) x) // convert everything to double
> > 		.map(x -> x - IntStream.of(data).average().getAsDouble()) // x - minus average
> > 		.map(x -> x * x) // this is to square the numerator
> > 		.reduce(0, (x,y) -> (x + (y/(data.length - 1))))); // sum everything together
> > }
> > ```
> 
> I don't think this method works when there are 0 elements in the data array. It will return OptionalDouble[0.0] instead of OptionalDouble.empty. If there is only 1 element in the data array, it will give out a OptionalDouble[NaN], which isn't ideal either. I think can use an if statement to catch these corner cases

Good catch @Buwoo! We can actually change the last reduce to `.reduce((x, y) -> (x + (y/(data.length - 1))))` to handle the corner cases. This way, we also don't need to do OptionalDouble.of(...).  The method will return a OptionalDouble.empty if there 0 elements in the data array. 

However, this does not handle the case when there is only 1 element in the data array. Probably should check this before running this.
